### PR TITLE
another round of image updates

### PR DESCRIPTION
**Pull request by benoit-pierre**
Highlights:
- update base image base image: there does not seem to be any advantage in using `phusion/baseimage:focal-1.2.0` over the official `ubuntu:focal`. The later is smaller, and the fact that the former README still mention upstart does not fill me with confidence.
- updated luajit & luarocks: while building the base image, I hit an issue installing some LUA packages: it looks like our luarocks version is not supported anymore (can't handle newer rockspecs, e.g.: `penlight-1.14.0-1.rockspec: Unknown field test`).
- updated clang image: downgrade clang to 14 for now (until we bump our glib version).
- stick to gcc 9: I gave up trying to install gcc 10 only, since we can't do without gcc 9, we might as well stick with it for a smaller image.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/virdevenv/94)

### CODE REVIEW

**Comment by Frenzie** [comment]
> * The later is smaller, and the fact that the former README still mention upstart does not fill me with confidence.

That's just a harmless outdated remark. It's been maintained all this time.

I do however agree that for this use case it doesn't seem to offer any advantages.

**Comment by Frenzie** [comment]
> - don't install koxtoolchain build dependencies
  (since we install pre-built releases)

I'm not necessarily sold on hyperoptimizing for a few stray kilobytes. Prebuilt releases are convenient to save time but I feel like there's a bit of a Catch 22 if you're throwing up hurdles to testing the non-existent new release.

File: docker/ubuntu/baseimage-20.04/Dockerfile [hunk 1]
```diff
@@ -88,41 +87,71 @@ RUN chmod +x local/bin/shfmt
...
-# LUAROCKS. {{{
+# LUAJIT. {{{
 
-RUN git clone https://github.com/torch/luajit-rocks.git
-WORKDIR luajit-rocks
-RUN git checkout 2c7496b905f6f972673effda4884766433b7583b
-RUN echo $HOME
-RUN cmake . -DWITH_LUAJIT21=ON -DCMAKE_INSTALL_PREFIX="$HOME/local"
-RUN make install
+ARG LUAJIT_VERSION=2.1.0-beta3
```
Comment (Frenzie) [inline docker/ubuntu/baseimage-20.04/Dockerfile:86]: It's worth noting that that is just as old as the one you're replacing. I'd probably just use the same commit hash as in https://github.com/koreader/koreader-base/pull/1765

File: docker/ubuntu/baseimage-20.04/Dockerfile [hunk 2]
```diff
@@ -6,6 +6,10 @@ FROM $REGISTRY/ubuntu:focal AS build
 # BOOTSTRAP {{{
 
 ARG DEBIAN_FRONTEND=noninteractive
+RUN tee /etc/apt/apt.conf.d/bare_installs.conf <<EOF
+APT::Install-Suggests "0";
+APT::Install-Recommends "0";
+EOF
```
Comment (Frenzie) [inline docker/ubuntu/baseimage-20.04/Dockerfile:12]: I find this reduces clarity, but I take it you're of a different opinion.

File: docker/ubuntu/koappimage/Dockerfile [hunk 3]
```diff
@@ -6,7 +6,8 @@ FROM base AS build
 USER root
 RUN apt-get update
 RUN apt-get upgrade -y
-RUN apt-get install -y --no-install-recommends libsndio7.0
+# NOTE: libx11-dev is needed to compile libxss-dummy.
+RUN apt-get install -y --no-install-recommends libsndio7.0 libx11-dev
```
[the remark this answers is not in the document]
Comment (Frenzie) [reply docker/ubuntu/koappimage/Dockerfile]: Ah, a following mini-commit.

File: docker/ubuntu/baseimage-20.04/Dockerfile (hunk 1 above)
Comment (benoit-pierre) [reply docker/ubuntu/baseimage-20.04/Dockerfile:86]: I had a hard time finding a combination of luajit/lua-lanes that does not result in a crashing and/or hanging luacheck. And no combo can pass the lua-lanes testsuite…

**Comment by benoit-pierre** [comment]
> > * don't install koxtoolchain build dependencies
> >   (since we install pre-built releases)
> 
> I'm not necessarily sold on hyperoptimizing for a few stray kilobytes.

Aggregated, the savings are pretty significant, e.g. for kobase: 315.52MB → 213.99MB for the compressed size (937MB → 630MB uncompressed).

> Prebuilt releases are convenient to save time but I feel like there's a bit of a Catch 22 if you're throwing up hurdles to testing the non-existent new release.

I don't understand.

File: docker/ubuntu/baseimage-20.04/Dockerfile (hunk 2 above)
Comment (benoit-pierre) [reply docker/ubuntu/baseimage-20.04/Dockerfile:12]: I got tired of typing it every time I was running the image…

**Comment by Frenzie** [comment]
> Aggregated, the savings are pretty significant, e.g. for kobase: 315.52MB → 213.99MB for the compressed size (937MB → 630MB uncompressed).

…huh? What are you talking about? There's no way ~10 MB uncompressed becomes a factor more.

> I don't understand.

As in frictionless building of a new toolchain in the image. The toolchain has to prove itself before it can become a release in the first place.

File: docker/ubuntu/baseimage-20.04/Dockerfile (hunk 2 above)
Comment (Frenzie) [reply docker/ubuntu/baseimage-20.04/Dockerfile:12]: Maybe just do both?

**Comment by benoit-pierre** [comment]
> > Aggregated, the savings are pretty significant, e.g. for kobase: 315.52MB → 213.99MB for the compressed size (937MB → 630MB uncompressed).
> 
> …huh? What are you talking about? There's no way ~10 MB uncompressed becomes hundreds of MB.

I'm saying that all those "small" optimizations — a few megabytes here, a dozen there — do add up to a 30% reduction in the final image size.

In that particular instance, it saves 22.5MB:
```
The following NEW packages will be installed:
  bison flex gawk gperf help2man liblocale-gettext-perl libncurses-dev libncurses5-dev libreadline8 libtext-unidecode-perl
  libxml-libxml-perl libxml-namespacesupport-perl libxml-sax-base-perl libxml-sax-perl readline-common tex-common texinfo ucf
0 upgraded, 18 newly installed, 0 to remove and 0 not upgraded.
Need to get 4172 kB of archives.
After this operation, 22.5 MB of additional disk space will be used.
```

> > I don't understand.
> 
> As in frictionless building of a new toolchain in the image. The toolchain has to prove itself before it can become a release in the first place.

Isn't that what the Github Actions in koxtoolchain are for? Considering how long a toolchain build is, having to install those dependencies has no meaningful impact on the total build time. But the reduced image size impact on all those CircleCI run adds up.

File: docker/ubuntu/baseimage-20.04/Dockerfile (hunk 2 above)
Comment (benoit-pierre) [reply docker/ubuntu/baseimage-20.04/Dockerfile:12]: That does not make sense to me.

I already use something like `apt() { sudo apt "$@"; }` when running an image, and it looks like I can change it to `apt() { sudo apt --no-install-recommends "$@"; }` and still run `apt update`, `apt remove`, etc…

So I'll remove the config.

**Comment by Frenzie** [comment]
> I'm saying that all those "small" optimizations — a few megabytes here, a dozen there — do add up to a 30% reduction in the final image size.

Optimizations and discarding goals of the image are different things. The former is a no-brainer, the latter is a decision.

> Considering how long a toolchain build is, having to install those dependencies has no meaningful impact on the total build time.

I said friction, not build time. But very well, we'll live with the friction. There's an apt-get copy/paste over on koxtoolchain.

**Pull request merged**